Move and update "memory model" docs (#301)
NVIDIA/libcudacxx: docs/extended_api/memory_model.md

enum thread_scope {
  thread_scope_system,
  thread_scope_device,
  thread_scope_block,
  thread_scope_thread
};

}  // namespace cuda
```

[`atomic`]: synchronization_primitives/atomic.md
[`barrier`]: synchronization_primitives/barrier.md

### Scope Relationships

Each program thread is related to each other program thread by one or more thread scope relations:
- Each thread in the system is related to each other thread in the system by the *system* thread scope: `thread_scope_system`.
- Each GPU thread is related to each other GPU thread in the same CUDA device by the *device* thread scope: `thread_scope_device`.
- Each GPU thread is related to each other GPU thread in the same CUDA thread block by the *block* thread scope: `thread_scope_block`.
- Each thread is related to itself by the `thread` thread scope: `thread_scope_thread`.

## Synchronization primitives

Types in namespaces `std::` and `cuda::std::` have the same behavior as corresponding types in namespace `cuda::` when instantiated with a scope of `cuda::thread_scope_system`.

## Atomicity

An atomic operation is atomic at the scope it specifies if:
- it specifies a scope other than `thread_scope_system`, **or**

> the scope is `thread_scope_system` and:

Reviewer: Why separate this part out?  Isn't this whole thing logically equivalent to just listing all 5 bullets in a single list?

Author: Yes, you are right.  The numbered list below says that this "boolean" expression is true:  1 || (2 && (3 || 4 || 5 || 6)). 
I got feedback that this was hard to follow when it is all together, so I decided to split it when the parenthesis open.

In [rendered](https://github.com/gonzalobg/libcudacxx/blob/docs/memory_model/docs/extended_api/memory_model.md#atomicity) markdown it looked ok to me, but if this can be improved I'd like to improve it.

An atomic operation is atomic at the scope it specifies if:
1. it specifies a scope other than `thread_scope_system`, or
2. the scope is `thread_scope_system` and
3. it affects an object in [unified memory] and [`concurrentManagedAccess`] is `1`, or
4. it affects an object in CPU memory and [`hostNativeAtomicSupported`] is `1`, or
5. it is a load or store that affects a naturally-aligned object of sizes `1`, `2`, `4`, or `8` bytes on [mapped memory], or
6. it affects an object in GPU memory and only GPU threads access it.

---
parent: Extended API
nav_order: 0
---

# Memory model

Standard C++ presents a view that the cost to synchronize threads is uniform and low.

CUDA C++ is different: the cost to synchronize threads grows as threads are further apart.
It is low across threads within a block, but high across arbitrary threads in the system running on multiple GPUs and CPUs.

To account for non-uniform thread synchronization costs that are not always low, CUDA C++ extends the standard C++ memory model and concurrency facilities in the `cuda::` namespace with **thread scopes**, retaining the syntax and semantics of standard C++ by default.

> ## Threads
> 
> In CUDA C++, memory can only be modified by threads, so any operation that modifies memory is being executed by a thread.
> 
> 
> [`memcpy_async`]: asynchronous_operations/memcpy_async.md
> [`cudaMemcpyAsync`]: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79
> 
> 

Author: Fix: 

```suggestion
```

---
parent: Extended API
nav_order: 0
---

# Memory model

Standard C++ presents a view that the cost to synchronize threads is uniform and low.

> CUDA C++ is different: the cost to synchronize threads grows as threads are further appart.

Reviewer: apart

Author: ```suggestion
CUDA C++ is different: the cost to synchronize threads grows as threads are further apart.
```

It is low across threads within a block, but high across arbitrary threads in the system running on multiple GPUs and CPUs.

To account for non-uniform thread synchronization costs that are not always low, CUDA C++ extends the standard C++ memory model and concurrency facilities in the `cuda::` namespace with **thread scopes**, retaining the syntax and semantics of standard C++ by default.

## Threads

In CUDA C++, memory can only be modified by threads, so any operation that modifies memory is being executed by a thread.

> Examples of threads of execution in CUDA C++ programs: GPU or CPU threads, threads in DMA engines executing asynchronous operations like [`memcpy_async`] or [`cudaMemcpyAsync`], threads in NICs executing RDMA operations, etc.

Reviewer: NACK this new Threads subsection.  We shouldn't talk about DMA engines, NICs, etc. here, and we shouldn't claim that nothing outside a thread can modify memory.

Author: This section does not claim "that nothing outside a thread can modify memory." but rather that those operations modify memory via a C++ thread.

I can modify this sentence to:

```suggestion
Examples of threads of execution in CUDA C++ programs: GPU threads, CPU threads, thread asynchronously copying memory for [`memcpy_async`] or [`cudaMemcpyAsync`], etc.
```

Author: Leaving sentence out: 

```suggestion
```

- it is a load or store that affects a naturally-aligned object of sizes `1`, `2`, `4`, or `8` bytes on [mapped memory], **or**
- it affects an object in GPU memory and only GPU threads access it.

[mapped memory]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#mapped-memory

Refer to the [CUDA programming guide] for more information on [unified memory], [mapped memory], CPU memory, and GPU peer memory.

[mapped memory]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#mapped-memory
[unified memory]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#um-unified-memory-programming-hd
[CUDA programming guide]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html
[`concurrentManagedAccess`]: https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html#structcudaDeviceProp_116f9619ccc85e93bc456b8c69c80e78b
[`hostNativeAtomicSupported`]: https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html#structcudaDeviceProp_1ef82fd7d1d0413c7d6f33287e5b6306f

## Data Races

Modify [intro.races paragraph 21] of ISO/IEC IS 14882 (the C++ Standard) as follows:

> > The execution of a program contains a data race if it contains two potentially concurrent conflicting actions, at least one of which is not atomic ***at a scope that includes the thread that performed the other operation***, and neither happens before the other, except for the special case for signal handlers described below. Any such data race results inundefined behavior. [...]

Reviewer: in undefined

Author: Fix typo: 
```suggestion
> The execution of a program contains a data race if it contains two potentially concurrent conflicting actions, at least one of which is not atomic ***at a scope that includes the thread that performed the other operation***, and neither happens before the other, except for the special case for signal handlers described below. Any such data race results in undefined behavior. [...]
```

<table class="display">
<tr class="header"><td colspan="2" markdown="span" align="center">
`int x = 0;`<br>
`int f = 0;`
</td></tr>
<tr class="header">
<td markdown="span" align="center"> 
**Thread 0 Block 0** 
</td><td markdown="span" align="center"> 
**Thread 0 Block 1** 
</td>
</tr>
<tr>
<td markdown="span">
`x = 42;`<br>
`cuda::atomic_ref<int, cuda::thread_scope_device> flag(f);`<br>
`flag.store(1, memory_order_release);`
</td>
<td markdown="span">
`cuda::atomic_ref<int, cuda::thread_scope_device> flag(f);`<br>
`while(flag.load(memory_order_acquire) != 1);`<br>
`assert(x == 42);`
</td>
</tr>
</table>

In the following variation of the previous example, two threads concurrently access the `f` object without synchronization, which leads to a **data race**, and exhibits **undefined behavior**:

<table>
<tr><td colspan="2" markdown="span" align="center">
`int x = 0;`<br>
`int f = 0;`
</td></tr>
<tr>
<td markdown="span" align="center"> 
**Thread 0 Block 0** 
</td><td markdown="span" align="center"> 
**Thread 0 Block 1** 
</td>
</tr>
<tr>
<td markdown="span">

> `x = 42;`<br>

Author: The documentation system that libcudacxx uses appears to have a bug that causes this.

Filled: https://github.com/NVIDIA/libcudacxx/issues/302
